### PR TITLE
Ecdc 2968 multi hits

### PR DESCRIPTION
**Pull request by JennyLouise**
### Issue reference / description

The branch you merge from should already reference an event-formation-unit github ticket number. You can add a descriptive title, but if an issue is referenced, you don't have to.

## Checklist for submitter

- [ ] Check for conflict with integration test
- [ ] Unit tests pass

---

## Nominate for Group Code Review (Anyone can nominate it)
Indicate if you think the code should be reviewed in a Thursday code review session.

- [ ] Recommend for group code review

Also, nominate it on the code_review Slack channel.

### CODE REVIEW

File: src/common/reduction/matching/GapMatcher.cpp [hunk 1]
```diff
@@ -63,3 +70,121 @@ std::string GapMatcher::config(const std::string &prepend) const {
...
+  } else {
+    return false;
+  }
+}
+
+void GapMatcher::check_and_stash_event(Event evt) {
+  if (!split_multi_events_) {
+    XTRACE(CLUSTER, DEB, "Stashing event");
+    stash_event(evt);
+    evt.clear();
```
Comment (JennyLouise) [inline src/common/reduction/matching/GapMatcher.cpp:169]: Does evt.clear() affect the original evt outside this function

File: src/common/reduction/matching/GapMatcher.cpp [hunk 2]
```diff
@@ -63,3 +70,121 @@ std::string GapMatcher::config(const std::string &prepend) const {
...
   return ss.str();
 }
+
+void GapMatcher::split_and_stash_event(Event evt) {
+  Cluster new_cluster;
+  std::vector<Cluster> new_clusters_a = split_cluster(evt.ClusterA);
+  std::vector<Cluster> new_clusters_b = split_cluster(evt.ClusterB);
+  std::vector<Event> new_events;
+
+  for (Cluster cluster_a : new_clusters_a) {
```
Comment (JennyLouise) [inline src/common/reduction/matching/GapMatcher.cpp:80]: TODO, ask francesco/find stats on the freia multiple events per time window, consider changing out to double hits instead of unlimited

File: src/common/reduction/matching/GapMatcher.cpp [hunk 3]
```diff
@@ -63,3 +70,121 @@ std::string GapMatcher::config(const std::string &prepend) const {
...
+    XTRACE(DATA, DEB,
+           "New cluster created, starts at coord %u, ends at coord %u",
+           cluster.coord_start(), cluster.coord_end());
+  }
+
+  return new_clusters;
+}
+
+bool GapMatcher::clusters_match(Cluster cluster_a, Cluster cluster_b) {
+  if ((cluster_a.weight_sum() * coefficient_ >=
```
Comment (JennyLouise) [inline src/common/reduction/matching/GapMatcher.cpp:155]: abs(a * coefficient - b) < allowance, check performance

File: src/common/reduction/matching/GapMatcher.cpp [hunk 4]
```diff
@@ -18,6 +18,13 @@ void GapMatcher::set_minimum_time_gap(uint64_t minimum_time_gap) {
   minimum_time_gap_ = minimum_time_gap;
 }
 
+void GapMatcher::set_split_multi_events(bool split_multi_events,
+                                        float coefficient, uint16_t allowance) {
+  split_multi_events_ = split_multi_events;
```
Comment (JennyLouise) [inline src/common/reduction/matching/GapMatcher.cpp:23]: Add counters

File: .gitignore [hunk 5]
```diff
@@ -16,6 +16,7 @@ a.out
 build*/
 cmake-build-*/
 doxygen/
+event-formation-unit/
```
Comment (mortenjc) [inline .gitignore:19]: Why is this needed?

File: src/common/reduction/matching/GapMatcher.cpp [hunk 6]
```diff
@@ -1,4 +1,4 @@
-/* Copyright (C) 2018 European Spallation Source, ERIC. See LICENSE file */
+/* Copyright (C) 2018-2022 European Spallation Source, ERIC. See LICENSE file */
```
Comment (mortenjc) [inline src/common/reduction/matching/GapMatcher.cpp:1]: Move to c++ comment while  editing
// Copyright (C) 2018 - 2022 European Spallation Source, ERIC. See LICENSE file

File: src/common/reduction/matching/GapMatcher.cpp [hunk 7]
```diff
@@ -63,3 +71,133 @@ std::string GapMatcher::config(const std::string &prepend) const {
...
+  bool filled_cluster_1 = false;
+
+  for (Hit hit : cluster.hits) {
+    // no hits added to any clusters yet, automatically added to cluster 1
+    if ((new_cluster_1->empty())) {
+      last_coord = hit.coordinate;
+      new_cluster_1->insert(hit);
+      XTRACE(CLUSTER, DEB, "Adding first hit to cluster 1");
+    }
+    // cluster 1 not marked as full yet, and gap is small enough to still be
```
Comment (mortenjc) [inline src/common/reduction/matching/GapMatcher.cpp:129]: somewhat bad commenting style to break up if else.

File: src/common/reduction/matching/GapMatcher.cpp [hunk 8]
```diff
@@ -63,3 +71,133 @@ std::string GapMatcher::config(const std::string &prepend) const {
...
+             "More than 2 distinct clusters in plane, clearing clusters and "
+             "returning");
+      new_cluster_1->clear();
+      new_cluster_2->clear();
+      return;
+    }
+  }
+}
+
+bool GapMatcher::clusters_match(Cluster cluster_a, Cluster cluster_b) {
```
Comment (mortenjc) [inline src/common/reduction/matching/GapMatcher.cpp:168]: might be good with an explanation here, also I think that probably some other code
would be better, like

abs(a-sum - b-sum) < tolerence (if I understood this correctly)

File: src/common/reduction/matching/GapMatcher.h [hunk 9]
```diff
@@ -1,4 +1,5 @@
-/* Copyright (C) 2018 European Spallation Source, ERIC. See LICENSE file */
+/* Copyright (C) 2018 - 2022 European Spallation Source, ERIC. See LICENSE file
```
Comment (mortenjc) [inline src/common/reduction/matching/GapMatcher.h:1]: Conver to C++ style comment we editing

File: src/common/reduction/matching/GapMatcher.h [hunk 10]
```diff
@@ -26,6 +27,9 @@ class GapMatcher : public AbstractMatcher {
   ///         the clusters are merged into one event.
   void set_minimum_time_gap(uint64_t minimum_time_gap);
 
+  void set_split_multi_events(bool split_multi_events, float coefficient_low,
+                              float coefficient_high);
+
   /// \brief Match queued up clusters into events.
   ///         Clusters that either overlap in time or have time gaps that are
```
Comment (mortenjc) [inline src/common/reduction/matching/GapMatcher.h:34]: un-tab text

File: src/common/reduction/matching/GapMatcher.h [hunk 11]
```diff
@@ -36,6 +40,30 @@ class GapMatcher : public AbstractMatcher {
   /// \brief print configuration of GapMatcher
   std::string config(const std::string &prepend) const override;
 
+  void reset_stats() {
```
Comment (mortenjc) [inline src/common/reduction/matching/GapMatcher.h:43]: better to use memset , then you don't need to remember adding anything here if you 
extend stats
memset(Stats, 0, sizeof(Stats))

File: src/common/reduction/matching/GapMatcher.h [hunk 12]
```diff
@@ -36,6 +40,30 @@ class GapMatcher : public AbstractMatcher {
...
+
+  struct Stats {
+    uint16_t SpanTooLarge{0};
+    uint16_t DiscardedSpanTooLarge{0};
+    uint16_t SplitSpanTooLarge{0};
+  } Stats;
+
 private:
+private:
+  void split_and_stash_event(Event evt);
```
Comment (mortenjc) [inline src/common/reduction/matching/GapMatcher.h:57]: I'd like to add doxygen comments, generally to all methods in .h files (I might forget to do this too, but I
think this is one of the most important things to do)
/// \brief ...
/// \param cluster ...
...
/// \return if method returns anything

File: src/common/reduction/matching/GapMatcher.h [hunk 13]
```diff
@@ -36,6 +40,30 @@ class GapMatcher : public AbstractMatcher {
...
+  void split_cluster(Cluster cluster, Cluster *new_cluster_1,
+                     Cluster *new_cluster_2);
+  bool clusters_match(Cluster cluster_a, Cluster cluster_b);
+  void check_and_stash_event(Event evt);
+
   uint64_t minimum_time_gap_{0};
```
Comment (mortenjc) [inline src/common/reduction/matching/GapMatcher.h:63]: consider putting these in a struct so show that they belong together

File: src/common/reduction/matching/test/GapMatcherTest.cpp [hunk 14]
```diff
@@ -1,13 +1,54 @@
-/** Copyright (C) 2017 European Spallation Source ERIC */
+/** Copyright (C) 2017-2022 European Spallation Source ERIC */
```
Comment (mortenjc) [inline src/common/reduction/matching/test/GapMatcherTest.cpp:1]: c++ comments but also add spaces between dash and the latest version that includes reference to LICENSE

File: src/modules/freia/FreiaBase.cpp [hunk 15]
```diff
@@ -211,6 +211,14 @@ void FreiaBase::processing_thread() {
 
       for (auto &builder : Freia.builders) {
         Freia.generateEvents(builder.Events);
+        if (Freia.Conf.SplitMultiEvents) {
+          Counters.EventsSpanTooLarge += builder.matcher.Stats.SpanTooLarge;
+          Counters.EventsDiscardedSpanTooLarge +=
+              builder.matcher.Stats.DiscardedSpanTooLarge;
+          Counters.EventsSplitSpanTooLarge +=
+              builder.matcher.Stats.SplitSpanTooLarge;
+          builder.matcher.reset_stats();
```
Comment (mortenjc) [inline src/modules/freia/FreiaBase.cpp:220]: I believe that the coding standard actually mandates the naming

resetStats()

File: src/modules/freia/geometry/Config.h [hunk 16]
```diff
@@ -42,6 +42,9 @@ class Config : public VMM3Config {
   bool WireGapCheck{true};
   uint16_t MaxGapWire{0};
   uint16_t MaxGapStrip{0};
+  bool SplitMultiEvents{false};
+  float SplitMultiEventsCoefficientLow{0.8};
```
Comment (mortenjc) [inline src/modules/freia/geometry/Config.h:46]: This should probably be a single tolerance value.

File: src/modules/nmx/NMXBase.cpp [hunk 17]
```diff
@@ -209,6 +209,14 @@ void NmxBase::processing_thread() {
 
       for (auto &builder : NMX.builders) {
         NMX.generateEvents(builder.Events);
+        if (NMX.Conf.NMXFileParameters.SplitMultiEvents) {
+          Counters.EventsSpanTooLarge += builder.matcher.Stats.SpanTooLarge;
```
Comment (mortenjc) [inline src/modules/nmx/NMXBase.cpp:213]: This can be done more elegantly with 

Counters.Matcher = builder.matcher.Stats if we keep the stats in a struct

File: src/common/reduction/matching/GapMatcher.cpp [hunk 18]
```diff
@@ -18,6 +18,14 @@ void GapMatcher::set_minimum_time_gap(uint64_t minimum_time_gap) {
   minimum_time_gap_ = minimum_time_gap;
 }
 
+void GapMatcher::set_split_multi_events(bool split_multi_events,
```
Comment (mortenjc) [inline src/common/reduction/matching/GapMatcher.cpp:21]: this is not the ECDC coding standard, should be
setSplitMultiEvents

depending on the coding style in this file

File: src/common/reduction/matching/GapMatcher.cpp [hunk 19]
```diff
@@ -63,3 +71,133 @@ std::string GapMatcher::config(const std::string &prepend) const {
...
+    //\todo count discarded multievents
+    return;
+  }
+  Stats.SplitSpanTooLarge++;
+  stash_event(new_event_1);
+  stash_event(new_event_2);
+  evt.clear();
+}
+
+void GapMatcher::split_cluster(Cluster cluster, Cluster *new_cluster_1,
```
Comment (mortenjc) [inline src/common/reduction/matching/GapMatcher.cpp:116]: Doxygen

File: src/common/reduction/matching/GapMatcher.h [hunk 20]
```diff
@@ -26,6 +27,9 @@ class GapMatcher : public AbstractMatcher {
   ///         the clusters are merged into one event.
   void set_minimum_time_gap(uint64_t minimum_time_gap);
 
+  void set_split_multi_events(bool split_multi_events, float coefficient_low,
```
Comment (mortenjc) [inline src/common/reduction/matching/GapMatcher.h:30]: All methods specified in .h files should have Doxygen comments describing what they to, what
parameters they take and, if relevant, the returned data.

I also forget to do this so we should keep reminding each other about this

File: src/common/reduction/matching/GapMatcher.cpp [hunk 21]
```diff
@@ -63,3 +71,133 @@ std::string GapMatcher::config(const std::string &prepend) const {
   ss << prepend << fmt::format("minimum_time_gap: {}\n", minimum_time_gap_);
   return ss.str();
 }
+
+void GapMatcher::split_and_stash_event(Event evt) {
```
Comment (mortenjc) [inline src/common/reduction/matching/GapMatcher.cpp:75]: This probably warrants some textual description

File: src/modules/nmx/geometry/Config.h [hunk 22]
```diff
@@ -45,6 +45,9 @@ class Config : public VMM3Config {
     uint16_t MaxXSpan{10};
     uint16_t MaxYSpan{10};
     uint16_t DefaultMinADC{50};
+    bool SplitMultiEvents{false};
+    float SplitMultiEventsCoefficientLow{0.8};
```
Comment (mortenjc) [inline src/modules/nmx/geometry/Config.h:49]: single tolerence parameter?

**Review by mortenjc — APPROVED** [review]
Nice work!
Most comments are about comments and documentation or style. This is slightly complicated so in some cases a bit more documentation would be good.

There are some suggestions for other implementations that are (in my opinion and if I undertand it coorrectly) more elegant or less error prone.

File: .gitignore (hunk 5 above)
Comment (JennyLouise) [reply .gitignore:19]: The performance tests run using the Centos artefact which is in a folder called event-formation-unit, so building the efu in a folder called "event-formation-unit" instead of "build" means you can run them locally without any changes to the script

File: src/modules/freia/geometry/Config.h (hunk 16 above)
Comment (JennyLouise) [reply src/modules/freia/geometry/Config.h:46]: I based this off Francesco's Freia multi-hit code where he has an option for a different low and high coefficient. There are also often situations where they won't centre around 1, where strips and wires have very different sensitivities, if I remember right the coefficients for freia are generally around 2.

File: src/common/reduction/matching/GapMatcher.cpp (hunk 6 above)
Comment (JennyLouise) [reply src/common/reduction/matching/GapMatcher.cpp:1]: Addressing in restyle branch

File: src/common/reduction/matching/GapMatcher.cpp (hunk 18 above)
Comment (JennyLouise) [reply src/common/reduction/matching/GapMatcher.cpp:21]: Addressing in restyle branch

File: src/modules/nmx/geometry/Config.h (hunk 22 above)
Comment (JennyLouise) [reply src/modules/nmx/geometry/Config.h:49]: Will bear in mind that this could be done more optimally - will wait on further requirements for multi-events before implementing anything, functionality might be entirely unnecessary
